targets/simple_switch_grpc/tests/test_grpc_counter.cpp

```diff
@@ -0,0 +1,182 @@
+    ClientContext context;
+    auto status = Write(&context, write_request, &write_response);
+    EXPECT_TRUE(status.ok());
+  }
+
+  // Add second table entry
+  exact->set_value(std::string("\x00\x00\x00\x05", 4));
+  *(update->mutable_entity()) = entity;
+  {
+    p4::WriteResponse write_response;
+    ClientContext context;
+    auto status = Write(&context, write_request, &write_response);
+    EXPECT_TRUE(status.ok());
+  }
+
+  // Send a packet matching first entry
+  p4::bm::PacketStreamRequest packet_request;
+  packet_request.set_device_id(device_id);
+  packet_request.set_port(1);
+  packet_request.set_packet(std::string(10, '\xab'));
+  {
+    ClientContext context;
+    auto stream = dataplane_stub->PacketStream(&context);
+    stream->Write(packet_request);
+    p4::bm::PacketStreamResponse packet_response;
+    stream->Read(&packet_response);
+    EXPECT_EQ(packet_request.device_id(), packet_response.device_id());
+    EXPECT_EQ(packet_request.port(), packet_response.port());
+    EXPECT_EQ(packet_request.packet(), packet_response.packet());
+    stream->WritesDone();
+    auto status = stream->Finish();
+    EXPECT_TRUE(status.ok());
+  }
+
+  // Send a packet matching second entry
+  packet_request.set_packet(std::string(5, '\xab'));
+  {
+    ClientContext context;
+    auto stream = dataplane_stub->PacketStream(&context);
+    stream->Write(packet_request);
+    p4::bm::PacketStreamResponse packet_response;
+    stream->Read(&packet_response);
+    EXPECT_EQ(packet_request.device_id(), packet_response.device_id());
+    EXPECT_EQ(packet_request.port(), packet_response.port());
+    EXPECT_EQ(packet_request.packet(), packet_response.packet());
+    stream->WritesDone();
+    auto status = stream->Finish();
+    EXPECT_TRUE(status.ok());
+  }
+
+  p4::ReadRequest read_request;
+  read_request.set_device_id(device_id);
+  auto read_entity = read_request.add_entities();
+  auto direct_counter_entry = read_entity->mutable_direct_counter_entry();
+  direct_counter_entry->set_counter_id(dc_id);
+  *(direct_counter_entry->mutable_table_entry()) = entity.table_entry();
+  // Read counter of second entry
+  {
+    ClientContext context;
+    std::unique_ptr<grpc::ClientReader<p4::ReadResponse> > reader(
+    p4runtime_stub->Read(&context, read_request));
+    p4::ReadResponse read_response;
+    reader->Read(&read_response);
+    auto status = reader->Finish();
+    EXPECT_TRUE(status.ok());
+    EXPECT_EQ(read_response.entities_size(), 1);
+    EXPECT_TRUE(read_response.entities(0).has_direct_counter_entry());
+    EXPECT_EQ(
+      read_response.entities(0).direct_counter_entry().data().packet_count(),
+      1);
+    EXPECT_EQ(
+      read_response.entities(0).direct_counter_entry().data().byte_count(),
+      5);
+  }
+
+  // Read counter of first entry
+  exact->set_value(std::string("\x00\x00\x00\x0a", 4));
+  *(direct_counter_entry->mutable_table_entry()) = entity.table_entry();
+  {
+    ClientContext context;
```

Comment:
define helper function rather than duplicate the code

Reply:
Fixed